Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -16,16 +16,86 @@
 
 #include "Firestore/core/src/firebase/firestore/remote/serializer.h"
 
-// TODO(rsgowman): These are (currently!) unnecessary includes. Adding for now
-// to ensure we can find nanopb's generated header files.
-#include "Firestore/Protos/nanopb/google/firestore/v1beta1/document.pb.h"
-#include "Firestore/Protos/nanopb/google/protobuf/timestamp.pb.h"
+#include <pb_decode.h>
+#include <pb_encode.h>
 
 namespace firebase {
 namespace firestore {
 namespace remote {
 
-Serializer::Serializer() {
+using firebase::firestore::model::FieldValue;
+
+Serializer::ValueWithType Serializer::EncodeFieldValue(
+    const FieldValue& field_value) {
+  Serializer::ValueWithType proto_value{
+      field_value.type(), google_firestore_v1beta1_Value_init_default};
+  switch (field_value.type()) {
+    case FieldValue::Type::Null:
+      proto_value.value.null_value = google_protobuf_NullValue_NULL_VALUE;
+      break;
+    default:
+      // TODO(rsgowman): implement the other types
+      abort();
+  }
+  return proto_value;
+}
+
+void Serializer::EncodeValueWithType(const ValueWithType& value,
+                                     uint8_t* out_bytes,
+                                     size_t* inout_bytes_length) {
+  bool status;
+  pb_ostream_t stream;
+  switch (value.type) {
+    case FieldValue::Type::Null:
+      stream = pb_ostream_from_buffer(out_bytes, *inout_bytes_length);
+      status = pb_encode_tag(&stream, PB_WT_VARINT,
+                             google_firestore_v1beta1_Value_null_value_tag);
+      if (!status) {
+        // TODO(rsgowman): figure out error handling
+        abort();
+      }
+
+      status = pb_encode_varint(&stream, value.value.null_value);
+      if (!status) {
+        // TODO(rsgowman): figure out error handling
+        abort();
+      }
+
+      *inout_bytes_length = stream.bytes_written;
+      break;
+
+    default:
+      // TODO(rsgowman): implement the other types
+      abort();
+  }
+}
+
+FieldValue Serializer::DecodeFieldValue(
+    const Serializer::ValueWithType& value_proto) {
+  switch (value_proto.type) {
+    case FieldValue::Type::Null:
+      return FieldValue::NullValue();
+    default:
+      // TODO(rsgowman): implement the other types
+      abort();
+  }
+}
+
+Serializer::ValueWithType Serializer::DecodeValueWithType(const uint8_t* bytes,
+                                                          size_t length) {
+  pb_istream_t stream = pb_istream_from_buffer(bytes, length);
+  pb_wire_type_t wire_type;
+  uint32_t tag;
+  bool eof;
+  bool status = pb_decode_tag(&stream, &wire_type, &tag, &eof);
+  if (!status || wire_type != PB_WT_VARINT ||
+      tag != google_firestore_v1beta1_Value_null_value_tag) {
```

Comment:
Might follow the same code structure as in other APIs of Serializer, where there is a switch with only null type dealt rather than combining with the check of status, which is type-independant.